drivers/soundwire/sysfs_slave_dpn.c

```diff
@@ -0,0 +1,348 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright(c) 2015-2020 Intel Corporation.
+
+#include <linux/device.h>
+#include <linux/mod_devicetable.h>
+#include <linux/slab.h>
+#include <linux/sysfs.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_type.h>
+#include "bus.h"
+#include "sysfs_local.h"
+
+/*
+ * dpN tree sysfs
+ *
+ */
+
+struct sdw_dpn_tree {
+	struct kobject **nodes;
+};
+
+struct dpn_attribute;
+
+struct dpn_attribute {
+	struct attribute	attr;
+	ssize_t (*show)(struct sdw_slave *slave, int N, int dir,
+			struct dpn_attribute *attr, char *buf);
+};
+
+static int get_port_N_dir(struct kobject *kobj, int *N, int *dir,
+			  struct sdw_slave **slv)
+{
+	struct device *dev = kobj_to_dev(kobj->parent);
+	char *nodename;
+	char *rest;
+	char *token;
+	ssize_t ret;
+
+	/* the node name is dp<N>_<sink/src> */
+	nodename = kstrdup(kobj->name, GFP_KERNEL);
+	if (!nodename)
+		return -ENOMEM;
+
+	rest = nodename;
+	token = nodename;
+
+	/* skip dp */
+	token = strsep(&rest, "p");
+	if (!token) {
+		ret = -EINVAL;
+		goto err;
+	}
+
+	/* extract N */
+	token = strsep(&rest, "_");
+	if (!token) {
+		ret = -EINVAL;
+		goto err;
+	}
+
+	ret = kstrtoint(token, 10, N);
```

Comment:
as mentioned above, non-digits are caught, but it's still a bit too flexible and forgiving for us. Firstly, I'd use `kstrtouint()` - negative numbers are invalid, right? Then `kstrtouint()` also allows a `+` sign in the beginning and a new line in the end. So, if we want to be bullet-safe we could check to exclude those too.

Reply:
no longer needed